src/cascadia/TerminalSettingsModel/WslDistroGenerator.cpp

```diff
@@ -136,3 +143,201 @@ std::vector<Profile> WslDistroGenerator::GenerateProfiles()
+        if (result == ERROR_NO_MORE_ITEMS)
+        {
+            break;
+        }
+
+        if (result == ERROR_SUCCESS &&
+            length == 38 &&
+            buffer[0] == L'{' &&
+            buffer[37] == L'}')
+        {
+            guidStrings.emplace_back(&buffer[0], length);
+        }
+    }
+
+    return true;
+}
+
+// Function Description:
+// - Get the list of all the names of all the WSL distros from the registry. If
+//   we fail to open any regkey for the GUID of a distro, we'll just skip it.
+//   Places the names of all the distros into the "names" param.
+// Arguments:
+// - wslRootKey: the HKEY for the Lxss node.
+// - guidStrings: A list of all the GUIDs of the installed distros
+// - names: a vector that receives all the names of the installed distros.
+// Return Value:
+// - false if the root key was invalid, else true.
+static bool getWslNames(const wil::unique_hkey& wslRootKey,
+                        std::vector<std::wstring>& guidStrings,
+                        std::vector<std::wstring>& names)
+{
+    if (!wslRootKey)
+    {
+        return false;
+    }
+    for (const auto& guid : guidStrings)
+    {
+        wil::unique_hkey distroKey{ openDistroKey(wslRootKey, guid) };
+        if (!distroKey)
+        {
+            continue;
+        }
+
+        std::wstring buffer;
+        auto result = wil::AdaptFixedSizeToAllocatedResult<std::wstring, 256>(buffer, [&](PWSTR value, size_t valueLength, size_t* valueLengthNeededWithNull) -> HRESULT {
+            auto length = static_cast<DWORD>(valueLength);
+            const auto status = RegQueryValueExW(distroKey.get(), RegKeyDistroName, 0, nullptr, reinterpret_cast<BYTE*>(value), &length);
+            // length will receive the number of bytes - convert to a number of
+            // wchar_t's. AdaptFixedSizeToAllocatedResult will resize buffer to
+            // valueLengthNeededWithNull
+            *valueLengthNeededWithNull = (length / sizeof(wchar_t));
+            // If you add one for another trailing null, then there'll actually
+            // be _two_ trailing nulls in the buffer.
+            return status == ERROR_MORE_DATA ? ERROR_SUCCESS : status;
+        });
+
+        if (result != ERROR_SUCCESS)
+        {
+            continue;
+        }
+        names.emplace_back(std::move(buffer));
+    }
+    return true;
+}
+
+// Method Description:
+// - Generate a list of profiles for each on the installed WSL distros. This
+//   will first try to read the installed distros from the registry. If that
+//   fails, we'll fall back to the legacy way of launching WSL.exe to read the
+//   distros from the commandline. Reading the registry is slightly more stable
+//   (see GH#7199, GH#9905), but it is certainly BODGY
+// Arguments:
+// - <none>
+// Return Value:
+// - A list of WSL profiles.
+std::vector<Profile> WslDistroGenerator::GenerateProfiles()
+{
+    static wil::unique_hkey wslRootKey{ openWslRegKey() };

```

Comment:
Interesting; this will be held for the lifetime of the application. That's intentional, I presume?

Reply:
Oh, uh, maybe not. I guess I thought it would be easier than looking it up each time, but you know, we _should_ close it when we're done with it.